src/scheduling.jl

```diff
@@ -82,33 +119,39 @@ function get_vertices_promises(vertices::Vector, G::MetaDiGraph)
     return promises
 end
 
-function get_deps_promises(vertex_id, map, G)
-    incoming_data = []
-    if haskey(map, vertex_id)
-        for src in map[vertex_id]
-            push!(incoming_data, get_prop(G, src, :res_data))
-        end
-    end
-    return incoming_data
+function get_in_promises(G, vertex_id)
+    return [get_prop(G, src, :res_data) for src in inneighbors(G, vertex_id)]
+end
+
+function get_out_promises(G, vertex_id)
+    return [get_prop(G, src, :res_data) for src in outneighbors(G, vertex_id)]
 end
 
 function schedule_graph(G::MetaDiGraph)
-    inc_e_src_map = get_ine_map(G)
+    data_vertices = MetaGraphs.filter_vertices(G, :type, "DataObject")
+    sorted_vertices = MetaGraphs.topological_sort(G)
 
-    for vertex_id in MetaGraphs.topological_sort(G)
-        incoming_data = get_deps_promises(vertex_id, inc_e_src_map, G)
-        set_prop!(G, vertex_id, :res_data, Dagger.@spawn AVAILABLE_TRANSFORMS[get_prop(G, vertex_id, :type)](name, graph_id, vertex_id, incoming_data...))
+    for data_id in data_vertices
+        size = get_prop(G, data_id, :size_average_B)
+        set_prop!(G, data_id, :res_data, DataObject(nothing, size))
+    end
+
+    Dagger.spawn_datadeps() do
+        for vertex_id in setdiff(sorted_vertices, data_vertices) 
+            incoming_data = get_in_promises(G, vertex_id)
+            outgoing_data = get_out_promises(G, vertex_id)
+            transform = get_transform(G, vertex_id)
+            N_inputs = length(incoming_data)
+            res = Dagger.@spawn transform(In.(incoming_data)..., Out.(outgoing_data)...; N_inputs)
```

Comment:
I feel the whole chain to execute algorithm is a bit too complicated.
In principle all the parameters from DAG could be extracted and "algorithm" created before anything is submitted to Dagger. I think at this point it may by clearer to replace some closures with callable structs
I'd replace `get_transform`, `AVAILABLE_TRANSFORMS` and `_algorithm` with something like this
```jl
struct MockupAlgorithm
    name::String
    runtime::Float64
    input_length::UInt
    MockupAlgorithm(graph::MetaDiGraph, vertex_id::Int) = begin
        runtime = get_prop(graph, vertex_id, :runtime_average_s)
        name = get_prop(graph, vertex_id, :node_id)
        inputs = length(inneighbors(graph, vertex_id))
        new(name, runtime, inputs)
    end
end

function (alg::MockupAlgorithm)(args...)

    function execute!(alg::MockupAlgorithm, inputs, outputs)
        println("Executing $(alg.name)")
        populate_data_object!.(outputs)
        sleep(alg.runtime)
    end

    inputs = args[1:alg.input_length]
    outputs = args[alg.input_length+1:end]
    execute!(alg, inputs, outputs)
end

function schedule_algorithm!(graph::MetaDiGraph, vertex_id::Int)
    incoming_data = get_in_promises(graph, vertex_id)
    outgoing_data = get_out_promises(graph, vertex_id)
    algorithm = MockupAlgorithm(graph, vertex_id)
    Dagger.@spawn algorithm(In.(incoming_data)..., Out.(outgoing_data)...)
end
```
and then
```jl
function schedule_graph(G::MetaDiGraph)
    data_vertices = MetaGraphs.filter_vertices(G, :type, "DataObject")
    sorted_vertices = MetaGraphs.topological_sort(G)

    for data_id in data_vertices
        size = get_prop(G, data_id, :size_average_B)
        set_prop!(G, data_id, :res_data, DataObject(nothing, size))
    end

    Dagger.spawn_datadeps() do
        for vertex_id in setdiff(sorted_vertices, data_vertices)
            res = schedule_algorithm!(G, vertex_id)
            set_prop!(G, vertex_id, :res_data, res)
        end
    end
end
```
Having the algorithms as  immutable callable structs might be useful later when we revise  the pipeline scheduling